src/services/chain.service.ts

```diff
@@ -0,0 +1,248 @@
+            const getMktCap = await lastValueFrom(this._httpService.get(`https://api-osmosis.imperator.co/tokens/v2/mcap`).pipe(map((response) => response.data)));
+
+            // We exclude LUM's mcap as we have our seperate service for that
+            return getMktCap
+                .filter((el) => this._assetSymbol.some((f) => f === el.symbol && el.symbol !== AssetSymbol.LUM))
+                .map((el) => ({
+                    total_value_usd: el.market_cap,
+                    symbol: el.symbol,
+                }));
+        } catch (error) {
+            this._logger.error(`Could not fetch Market Cap for External Chains...`, error);
+
+            Sentry.captureException(error);
+
+            return null;
+        }
+    };
+
+    getTokenSupply = async (): Promise<{ supply: number; symbol: string }[]> => {
+        try {
+            // Not all chains have a reliable token supply that we can extract from chain
+            // Hence for some, we need to rely on their official endpoint to retrieve it
+
+            // Return token supply based on the microDenum and denum from assets in the index
+            const chainSupply = await Promise.all(
+                this._client.map(async (el, index) => {
+                    const supply = Number((await el.getSupply(this._assetMicroDenum[index])).amount) / TEN_EXPONENT_SIX;
+
+                    return { supply, symbol: this._assetSymbol[index] };
+                }),
+            );
+
+            // The value returned from Evmos on chain does not seem accurate to us.
+            // Hence, we rely on their official documentation endpoint to retrieve the supply
+            const evmosSupply =
+                Number(
+                    await lastValueFrom(this._httpService.get(`https://rest.bd.evmos.org:1317/evmos/inflation/v1/circulating_supply`).pipe(map((response) => response.data.circulating_supply.amount))),
+                ) / CLIENT_PRECISION;
+
+            // We map the token supply from the other chains with the one from evmos
+            return chainSupply.map((el) => [{ supply: evmosSupply, symbol: AssetSymbol.EVMOS }].find((o) => o.symbol === el.symbol) || el);
+        } catch (error) {
+            this._logger.error(`Could not fetch Token Supply for External Chains...`, error);
+
+            Sentry.captureException(error);
+
+            return null;
+        }
+    };
+
+    getApy = async (): Promise<{ apy: number; symbol: string }[]> => {
+        try {
+            // As inflation is not retrievable for now for some chain
+            // We take the first 5 {atom, akt, cmdx, dvpn, ki} chain as we need to compute the last ones manually for now.
+            // Once we can compute the inflation of all chains evenly we integrate them in the getChainApy variable
+            const client = this._client.slice(0, this._dfractOnChainApy.length);
+
+            // Chains where inflation is retrievable
+            const getChainApy = await Promise.all(
```

Comment:
check promise

Reply:
Only place with chainSupply where we need this redundancy.